core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1531 @@
+   */
+  sealed case class MissingLogDirMoveState(targetLogDir: String)
+      extends ReplicaMoveState {
+    override def done = false
+  }
+
+  /**
+   * A replica move state where the source replica is missing.
+   *
+   * @param targetLogDir        The log directory that we wanted the replica to move to.
+   */
+  sealed case class MissingReplicaMoveState(targetLogDir: String)
+      extends ReplicaMoveState {
+    override def done = false
+  }
+
+  /**
+   * A replica move state where the move is in progress.
+   *
+   * @param currentLogDir       The current log directory.
+   * @param futureLogDir        The log directory that the replica is moving to.
+   * @param targetLogDir        The log directory that we wanted the replica to move to.
+   */
+  sealed case class ActiveMoveState(currentLogDir: String,
+                                    targetLogDir: String,
+                                    futureLogDir: String)
+      extends ReplicaMoveState {
+    override def done = false
+  }
+
+  /**
+   * A replica move state where there is no move in progress, but we did not
+   * reach the target log directory.
+   *
+   * @param currentLogDir       The current log directory.
+   * @param targetLogDir        The log directory that we wanted the replica to move to.
+   */
+  sealed case class CancelledMoveState(currentLogDir: String,
+                                       targetLogDir: String)
+      extends ReplicaMoveState {
+    override def done = true
+  }
+
+  /**
+   * The completed replica move state.
+   *
+   * @param targetLogDir        The log directory that we wanted the replica to move to.
+   */
+  sealed case class CompletedMoveState(targetLogDir: String)
+      extends ReplicaMoveState {
+    override def done = true
+  }
+
+  /**
+   * An exception thrown to indicate that the command has failed, but we don't want to
+   * print a stack trace.
+   *
+   * @param message     The message to print out before exiting.  A stack trace will not
+   *                    be printed.
+   */
+  class TerseReassignmentFailureException(message: String) extends KafkaException(message) {
+  }
 
   def main(args: Array[String]): Unit = {
     val opts = validateAndParseArgs(args)
-    val zkConnect = opts.options.valueOf(opts.zkConnectOpt)
-    val time = Time.SYSTEM
-    val zkClient = KafkaZkClient(zkConnect, JaasUtils.isZkSaslEnabled, 30000, 30000, Int.MaxValue, time)
-
-    val adminClientOpt = createAdminClient(opts)
+    var toClose: Option[AutoCloseable] = None
+    var failed = true
 
     try {
-      if(opts.options.has(opts.verifyOpt))
-        verifyAssignment(zkClient, adminClientOpt, opts)
-      else if(opts.options.has(opts.generateOpt))
-        generateAssignment(zkClient, opts)
-      else if (opts.options.has(opts.executeOpt))
-        executeAssignment(zkClient, adminClientOpt, opts)
+      if (opts.options.has(opts.bootstrapServerOpt)) {
```

Comment:
Maybe we could split this into two "main" methods: one for AdminClient and one for ZK? Might make this a bit more readable.

Reply:
ok